File: first_missing_positive.go
func firstMissingPositive(nums []int) int {
    for index, val := range nums {
        for val != index+1 && val > 0 && val <= len(nums) && val != nums[val-1] {
            nums[index], nums[val-1] = nums[val-1], nums[index]
            val = nums[index]
        }
    }

    for index, val := range nums {
        if val != index+1 {
            return index+1
        }
    }
    return len(nums) + 1
}


// Time Complexity : O(n)
// Space Complexity : O(1)

File: first_missing_positive.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        for(int i=0; i < nums.size(); i++) {
            if(nums[i] == i+1 || nums[i] < 1 || nums[i] > nums.size() || nums[nums[i]-1] == nums[i]) {
                continue;
            }
            swap(nums[i], nums[nums[i--]-1]);
        }

        int i;
        for(i=0; i < nums.size(); i++) {
            if(nums[i] != i+1) {
                return i+1;
            }
        }
        return i+1;
    }
};

// Time Complexity : O(n)
// Space Complexity : O(1)
